packages/cipherstash-proxy/src/cli/migrate/mod.rs

```diff
@@ -0,0 +1,309 @@
+    /// Name of database table
+    ///
+    #[arg(short, long)]
+    table: String,
+
+    ///
+    /// Source and destination columns as space-delimited key pairs `--columns source=destination`
+    ///
+    #[arg(short, required = true, long, num_args(1..), value_parser = parse_key_val::<String, String>)]
+    columns: Vec<(String, String)>,
+
+    ///
+    /// Primary key column/s
+    /// Compound primary keys can be provided as a space delimted list: `--primary-key id user_id`
+    ///
+    #[arg(short = 'k', long, num_args(1..), value_delimiter = ' ', default_values_t = vec![ID.to_string()])]
+    primary_key: Vec<String>,
+
+    // Updates `batch_size` records at a time
+    #[arg(short, long, default_value_t = 100)]
+    batch_size: usize,
+
+    /// Run without update. Data is fetched, but updates are not performed.
+    #[arg(short, long, default_value_t = false)]
+    dry_run: bool,
+
+    /// Turn on additional logging output
+    #[arg(short, long, default_value_t = false)]
+    verbose: bool,
+}
+
+impl Migrate {
+    ///
+    /// Returns true if this is not a `dry_run`
+    ///
+    fn commit(&self) -> bool {
+        !self.dry_run
+    }
+
+    ///
+    /// Run the encryption migration process
+    ///
+    pub async fn run(&self, config: TandemConfig) -> Result<(), Error> {
+        let connection_string = format!(
+            "postgresql://{}:{}@{}:{}/{}",
+            config.database.username,
+            config.database.password,
+            config.server.host,
+            config.server.port,
+            config.database.name
+        );
+
+        debug!(target: MIGRATE, ?config);
+
+        let client =
+            connect_with_tls(&connection_string, config.database.with_tls_verification).await?;
+
+        info!(
+            target: MIGRATE,
+            msg = "Connected to Database",
+            database = config.database.name,
+            host = config.server.host,
+            port = config.server.port,
+            username = config.database.username,
+            with_tls_verification = config.database.with_tls_verification,
+        );
+
+        info!(target: MIGRATE, msg = "Encrypting table", table = self.table, columns = ?self.columns);
+
+        if !self.commit() {
+            warn!(msg = "Dry run is enabled");
+        }
+
+        let mut batch_count = 0;
+        let mut updated_count = 0;
+
+        let quoted_table = escape_identifier(&self.table);
+
+        let primary_key_idents = self
+            .primary_key
+            .iter()
+            .map(|pk| escape_identifier(pk))
+            .collect::<Vec<String>>()
+            .join(", ");
+
+        let column_idents = self
+            .columns
+            .iter()
+            .map(|(source_col, _target_col)| escape_identifier(source_col))
+            .collect::<Vec<String>>()
+            .join(", ");
+
+        let sql =
+        format!("SELECT {primary_key_idents}, {column_idents} FROM {quoted_table} ORDER BY {primary_key_idents}");
```

Comment:
Is this safe? I'd prefer to use params over this

Reply:
Is a valid point. 
Will raise as a separate issue, as this is all existing code.
Everything does get quoted appropriately, using the same quoting from the rust-postgres tools that are used to escape params.

Reply:
Cool. I'm happy with addressing this in a separate issue.